File: services/academic-service/src/test/java/com/visor/school/academicservice/integration/StudentEnrollmentIntegrationTest.java
package com.visor.school.academicservice.integration;

import com.visor.school.academicservice.config.TestConfig;
import com.visor.school.academicservice.model.Address;
import com.visor.school.academicservice.model.EnrollmentStatus;
import com.visor.school.academicservice.model.Student;
import com.visor.school.academicservice.repository.StudentRepository;
import com.visor.school.academicservice.service.StudentService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@ActiveProfiles("test")
@Import(TestConfig.class)
@Transactional
class StudentEnrollmentIntegrationTest {

    @Autowired
    private StudentService studentService;

    @Autowired
    private StudentRepository studentRepository;

    @Test
    void shouldCompleteStudentEnrollmentFlow() {
        UUID userId = UUID.randomUUID();
        String firstName = "Integration";
        String lastName = "Test";
        LocalDate dateOfBirth = LocalDate.of(2010, 1, 1);
        int gradeLevel = 5;

        Student student = studentService.enrollStudent(userId, firstName, lastName, dateOfBirth, gradeLevel, null, null);

        assertNotNull(student);
        assertNotNull(student.getId());
        assertNotNull(student.getStudentId());
        assertFalse(student.getStudentId().isBlank());
        assertEquals(firstName, student.getFirstName());
        assertEquals(lastName, student.getLastName());
        assertEquals(dateOfBirth, student.getDateOfBirth());
        assertEquals(gradeLevel, student.getGradeLevel());
        assertEquals(EnrollmentStatus.ENROLLED, student.getEnrollmentStatus());

        var savedStudent = studentRepository.findById(student.getId());
        assertTrue(savedStudent.isPresent());
        assertEquals(firstName, savedStudent.get().getFirstName());
    }

    @Test
    void shouldEnrollStudentWithAddressAndEmergencyContact() {
        UUID userId = UUID.randomUUID();
        Address address = new Address("456 Test St", "Test City", "Test State", "12345", "Test Country");

        Student student = studentService.enrollStudent(userId, "With", "Contact", LocalDate.of(2010, 1, 1), 5, address, null);

        assertNotNull(student.getAddress());
        assertEquals("Test City", student.getAddress().getCity());
    }

    @Test
    void shouldFailEnrollmentWhenGradeLevelIsInvalid() {
        UUID userId = UUID.randomUUID();
        assertThrows(IllegalArgumentException.class, () ->
                studentService.enrollStudent(userId, "Invalid", "Grade", LocalDate.of(2010, 1, 1), 13, null, null)
        );
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/integration/StudentEnrollmentIntegrationTest.kt
package com.visor.school.academicservice.integration

import com.visor.school.academicservice.config.TestConfig
import com.visor.school.academicservice.model.Address
import com.visor.school.academicservice.model.EnrollmentStatus
import com.visor.school.academicservice.repository.StudentRepository
import com.visor.school.academicservice.service.StudentService
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.util.UUID
import com.visor.school.academicservice.model.EmergencyContact
import org.junit.jupiter.api.assertThrows

/**
 * Integration test for student enrollment flow
 */
@SpringBootTest
@ActiveProfiles("test")
@Import(TestConfig::class)
@Transactional
class StudentEnrollmentIntegrationTest @Autowired constructor(
    private val studentService: StudentService,
    private val studentRepository: StudentRepository
) {

    @Test
    fun `should complete student enrollment flow`() {
        // Given
        val userId = UUID.randomUUID()
        val firstName = "Integration"
        val lastName = "Test"
        val dateOfBirth = LocalDate.of(2010, 1, 1)
        val gradeLevel = 5

        // When
        val student = studentService.enrollStudent(
            userId = userId,
            firstName = firstName,
            lastName = lastName,
            dateOfBirth = dateOfBirth,
            gradeLevel = gradeLevel
        )

        // Then
        assertNotNull(student)
        assertNotNull(student.id!!)
        assertNotNull(student.studentId)
        assertFalse(student.studentId.isBlank())
        assertEquals(firstName, student.firstName)
        assertEquals(lastName, student.lastName)
        assertEquals(dateOfBirth, student.dateOfBirth)
        assertEquals(gradeLevel, student.gradeLevel)
        assertEquals(EnrollmentStatus.ENROLLED, student.enrollmentStatus)

        // Verify student is persisted
        val savedStudent = studentRepository.findById(student.id!!)
        assertTrue(savedStudent.isPresent)
        assertEquals(firstName, savedStudent.get().firstName)
    }

    @Test
    fun `should enroll student with address and emergency contact`() {
        // Given
        val userId = UUID.randomUUID()
        val address = Address(
            street = "456 Test St",
            city = "Test City",
            state = "Test State",
            postalCode = "12345",
            country = "Test Country"
        )

        // When
        val student = studentService.enrollStudent(
            userId = userId,
            firstName = "With",
            lastName = "Contact",
            dateOfBirth = LocalDate.of(2010, 1, 1),
            gradeLevel = 5,
            address = address
        )

        // Then
        assertNotNull(student.address)
        assertEquals("Test City", student.address?.city)
    }

    @Test
    fun `should fail enrollment when grade level is invalid`() {
        // Given
        val userId = UUID.randomUUID()

        // When & Then
        assertThrows<IllegalArgumentException> {
            studentService.enrollStudent(
                userId = userId,
                firstName = "Invalid",
                lastName = "Grade",
                dateOfBirth = LocalDate.of(2010, 1, 1),
                gradeLevel = 13
            )
        }
    }

    @Test
    fun `should fail enrollment when student already exists`() {
        // Given
        val userId = UUID.randomUUID()
        studentService.enrollStudent(
            userId = userId,
            firstName = "First",
            lastName = "Enrollment",
            dateOfBirth = LocalDate.of(2010, 1, 1),
            gradeLevel = 5
        )

        // When & Then
        assertThrows<IllegalArgumentException> {
            studentService.enrollStudent(
                userId = userId,
                firstName = "Second",
                lastName = "Enrollment",
                dateOfBirth = LocalDate.of(2010, 1, 1),
                gradeLevel = 5
            )
        }
    }

    @Test
    fun `should generate unique student IDs`() {
        // Given
        val userId1 = UUID.randomUUID()
        val userId2 = UUID.randomUUID()

        // When
        val student1 = studentService.enrollStudent(
            userId = userId1,
            firstName = "Student",
            lastName = "One",
            dateOfBirth = LocalDate.of(2010, 1, 1),
            gradeLevel = 5
        )
        val student2 = studentService.enrollStudent(
            userId = userId2,
            firstName = "Student",
            lastName = "Two",
            dateOfBirth = LocalDate.of(2010, 1, 1),
            gradeLevel = 5
        )

        // Then
        assertNotEquals(student1.studentId, student2.studentId)
        assertNotNull(student1.studentId)
        assertNotNull(student2.studentId)
    }
}
